src/signTxPoolRegistration.c

```diff
@@ -101,235 +125,696 @@ static inline void advanceState()
+	{
+		// sanity checks
+		CHECK_STATE(STAKE_POOL_REGISTRATION_REWARD_ACCOUNT);
 
-	security_policy_t policy = policyForSignTxCertificateStakePoolRegistration();
+		ASSERT(wireDataSize < BUFFER_SIZE_PARANOIA);
+	}
+	{
+		// parse data
+
+		TRACE_BUFFER(wireDataBuffer, wireDataSize);
+
+		read_view_t view = make_read_view(wireDataBuffer, wireDataBuffer + wireDataSize);
+
+		_parsePoolRewardAccount(&view);
+
+		VALIDATE(view_remainingSize(&view) == 0, ERR_INVALID_DATA);
+	}
+
+	security_policy_t policy = policyForSignTxStakePoolRegistrationRewardAccount(
+	                                   commonTxData->signTxUsecase,
+	                                   &subctx->stateData.poolRewardAccount
+	                           );
 	TRACE("Policy: %d", (int) policy);
 	ENSURE_NOT_DENIED(policy);
+
+	{
+		// key derivation must not be done before DENY security policy is enforced
+		uint8_t rewardAccountBuffer[REWARD_ACCOUNT_SIZE];
+		_calculateRewardAccount(&subctx->stateData.poolRewardAccount, rewardAccountBuffer);
+
+		txHashBuilder_poolRegistrationCertificate_rewardAccount(
+		        txHashBuilder,
+		        rewardAccountBuffer, SIZEOF(rewardAccountBuffer)
+		);
+	}
+
 	{
 		// select UI steps
 		switch (policy) {
 #	define  CASE(POLICY, UI_STEP) case POLICY: {subctx->ui_step=UI_STEP; break;}
-			CASE(POLICY_PROMPT_BEFORE_RESPONSE, HANDLE_POOLPARAMS_STEP_DISPLAY_OPERATION);
-			CASE(POLICY_ALLOW_WITHOUT_PROMPT,   HANDLE_POOLPARAMS_STEP_RESPOND);
+			CASE(POLICY_SHOW_BEFORE_RESPONSE, HANDLE_POOL_REWARD_ACCOUNT_STEP_DISPLAY);
+			CASE(POLICY_ALLOW_WITHOUT_PROMPT, HANDLE_POOL_REWARD_ACCOUNT_STEP_RESPOND);
 #	undef   CASE
 		default:
 			THROW(ERR_NOT_IMPLEMENTED);
 		}
 	}
 
-	handlePoolParams_ui_runStep();
+	handlePoolRewardAccount_ui_runStep();
 }
 
 // ============================== OWNER ==============================
 
 enum {
-	HANDLE_OWNER_STEP_DISPLAY = 6320,
+	HANDLE_OWNER_STEP_DISPLAY = 6600,
 	HANDLE_OWNER_STEP_RESPOND,
 	HANDLE_OWNER_STEP_INVALID,
 };
 
 static void handleOwner_ui_runStep()
 {
 	TRACE("UI step %d", subctx->ui_step);
+	TRACE_STACK_USAGE();
 	ui_callback_fn_t* this_fn = handleOwner_ui_runStep;
 
 	UI_STEP_BEGIN(subctx->ui_step);
 
 	UI_STEP(HANDLE_OWNER_STEP_DISPLAY) {
-		ui_displayPoolOwnerScreen(&subctx->owner, subctx->currentOwner, commonTxData->networkId, this_fn);
+		ui_displayPoolOwnerScreen(&subctx->stateData.owner, subctx->currentOwner, commonTxData->networkId, this_fn);
 	}
 	UI_STEP(HANDLE_OWNER_STEP_RESPOND) {
 		respondSuccessEmptyMsg();
 
 		subctx->currentOwner++;
 		if (subctx->currentOwner == subctx->numOwners) {
-			VALIDATE(subctx->numOwnersGivenByPath == 1, ERR_INVALID_DATA);
+			switch (commonTxData->signTxUsecase) {
+			case SIGN_TX_USECASE_POOL_REGISTRATION_OWNER:
+				VALIDATE(subctx->numOwnersGivenByPath == 1, ERR_INVALID_DATA);
+				break;
+
+				#ifdef POOL_OPERATOR_APP
+			case SIGN_TX_USECASE_POOL_REGISTRATION_OPERATOR:
+				VALIDATE(subctx->numOwnersGivenByPath <= 1, ERR_INVALID_DATA);
```

Comment:
do we want to allow to the operator to sign as an owner at the same time?

Reply:
it also seems to be in discrepancy with maxNumWitnesses which currently are taking into account only the amount of inputs and withdrawals (which I challenged if we actually need/want)

Reply:
ok it isn't in discrepancy, I just noticed the additional "2" being summed into the equation

Reply:
a possible problem I see with allowing signing a pool as an owner while in "operator mode" is that an attacker may unknowingly "coerce" an operator into signing up as an owner by requesting a witness for the operator's staking key which would be listed - the operator may or may not notice while checking the hashes that one of them is indeed their own - probably not a big deal but I would not mix the concerns unless really needed to prevent any such weird "loopholes"